### ISSUE TITLE
Bug with control-flow merge and null assignment

### ISSUE THREAD

**Issue by msridhar** [kelloggm/object-construction-checker#170]
Test case:
```java
class SimpleInferenceMerge {
    void build(@CalledMethods({"a", "b"}) SimpleInferenceMerge this) { }
    void a() {}
    void b() {}
    static void mergeWithNull(boolean b) {
        SimpleInferenceMerge s = null;
        if (b) {
            s = new SimpleInferenceMerge();
            s.a();
            s.b();
        }
        // :: error: finalizer.invocation.invalid
        s.build();
    }
}
```
The checker does not report the expected error at the `build()` call.  It seems the inferred type of `s` at the call is `@CalledMethods({"a", "b"})` which is clearly incorrect.

@kelloggm any idea what's going on here?

/cc @Nargeshdb who debugged this a bit

The test case is checked in on the `merge-bug` branch

**Comment by msridhar** [comment]
It seems that we assign `null` the type `@CalledMethodsBottom` which seems problematic, as that is equivalent to saying all possible methods have been called.  Maybe `@CalledMethodsTop` should be the type for `null` literals?

**Comment by kelloggm** [comment]
Sorry this is causing confusion!

The checker is behaving correctly. As @msridhar noted, we assign `null` the type `@CalledMethodsBottom`, following the standard Checker Framework practice (https://checkerframework.org/manual/#creating-bottom-and-top-qualifier).

To get an intuition for why the checker behaves this way, consider the question of whether there is any way for the program to execute in such a way that `build` is called on `s` in a state where `a()` and `b()` have not been called on `s` beforehand. The answer is no:
* if the boolean `b` is `true`, then `a()` and `b()` have been called on `s`, which is non-null. The call to `build` is legal.
* if the boolean `b` is `false`, then the program will crash with a null-pointer exception as soon as `s` is dereferenced, so `build` will never be called.

This subtlety gets to the separation of concerns that pluggable typechecking relies on - it is not the object construction checker's responsibility to ensure that all dereferences are legal, so it assumes that they are. A programmer concerned about null dereferences should run the nullness checker, or NullAway, or some other nullness analysis, which would permit them to catch this error. In a sense, this is a form of rely-guarantee reasoning: the object construction checker (and other CF checkers) *rely* on the fact that dereference is legal, because otherwise some other mechanism (the nullness checker or an NPE) will prevent whatever they are checking for from being relevant.

**Comment by msridhar** [comment]
Thanks, Martin.  Your explanation makes sense.  Basically the OCC gives guarantees under the condition that the calls actually succeed, rather than failing due to an NPE.
